File: examples/csharp/Scripts/LoopListScene.cs
using Godot;

namespace FairyGUI.Examples;

public partial class LoopListScene : DemoSceneBase
{
    private GuiObject _list;

    protected override void ContinueInit()
    {
        UIPackage.AddPackage("res://Resources/UI/LoopList");
        _view = UIPackage.CreateObject("LoopList", "Main");
        _groot.AddChild(_view);

        _list = _view.GetChild("list");
        if (_list == null)
            return;

        _list.SetItemRenderer(new Callable(this, MethodName.RenderListItem));
        _list.SetVirtualAndLoop();
        _list.SetNumItems(5);
        _list.AddEventListener(FguiEvent.Scroll, new Callable(this, MethodName.OnScroll));
        UpdateSpecialEffect();
    }

    private void RenderListItem(int index, GodotObject obj)
    {
        obj.SetPivot(0.5f, 0.5f, false);
        obj.SetIcon("ui://LoopList/n" + (index + 1));
    }

    private void OnScroll(FguiEventContext ctx)
    {
        UpdateSpecialEffect();
    }

    private void UpdateSpecialEffect()
    {
        if (_list == null)
            return;
        var sp = _list.GetScrollPane();
        if (sp == null)
            return;

        var midX = sp.GetScrollingPosX() + _list.GetViewWidth() / 2.0f;
        var cnt = _list.NumChildren();
        for (var i = 0; i < cnt; i++)
        {
            var obj = _list.GetChildAt(i);
            if (obj == null)
                continue;
            var dist = Mathf.Abs(midX - obj.GetX() - obj.GetWidth() / 2.0f);
            if (dist > obj.GetWidth())
            {
                obj.SetScaleX(1);
                obj.SetScaleY(1);
            }
            else
            {
                var ss = 1 + (1 - dist / obj.GetWidth()) * 0.24f;
                obj.SetScaleX(ss);
                obj.SetScaleY(ss);
            }
        }

        var n3 = _view.GetChild("n3");
        if (n3 != null)
        {
            var firstIdx = _list.GetFirstChildInView();
            var total = _list.GetNumItems();
            n3.SetText(((firstIdx + 1) % total).ToString());
        }
    }
}

File: examples/ts/Scripts/LoopListScene.ts
/// <reference path="../fairygui.d.ts" />

import { Callable } from "godot";
import { DemoSceneBase } from "./DemoSceneBase";

export default class LoopListScene extends DemoSceneBase {
    private _list: GList | null = null;

    protected ContinueInit(): void {
        UIPackage.addPackage("res://Resources/UI/LoopList");
        this._view = UIPackage.createObject("LoopList", "Main");
        this._groot!.addChild(this._view);

        this._list = this._view.getChild("list") as GList | null;
        if (this._list != null) {
            this._list.setItemRenderer(Callable.create(this._renderListItem.bind(this)));
            this._list.setVirtualAndLoop();
            this._list.setNumItems(5);
            this._list.addEventListener(
                UIEventDispatcher.SCROLL,
                Callable.create(this._doSpecialEffect.bind(this)),
            );
            this._doSpecialEffect();
        }
    }

    private _renderListItem(index: number, obj: GComponent): void {
        obj.setPivot(0.5, 0.5, false);
        obj.icon = `ui://LoopList/n${String(index + 1)}`;
    }

    private _doSpecialEffect(): void {
        if (this._list == null) {
            return;
        }
        const sp = this._list.getScrollPane();
        if (sp == null) {
            return;
        }
        const midX = sp.getPosX() + this._list.getViewWidth() / 2.0;
        const cnt = this._list.numChildren();
        for (let i = 0; i < cnt; i++) {
            const obj = this._list.getChildAt(i);
            if (obj == null) {
                continue;
            }
            const dist = Math.abs(midX - obj.getX() - obj.getWidth() / 2.0);
            if (dist > obj.getWidth()) {
                obj.setScaleX(1);
                obj.setScaleY(1);
            } else {
                const ss = 1 + (1 - dist / obj.getWidth()) * 0.24;
                obj.setScaleX(ss);
                obj.setScaleY(ss);
            }
        }

        const n3 = this._view!.getChild("n3");
        if (n3 != null) {
            const firstIdx = this._list.getFirstChildInView();
            const total = this._list.getNumItems();
            n3.setText(String((firstIdx + 1) % total));
        }
    }
}
